File: LeetCode/98验证二叉搜索树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
// bool judge(struct TreeNode *node, long a, long b) {
//     if (node == NULL) return true;
//     if (node->val <= a || node->val >= b) return false;
//     return judge(node->left, a, node->val) && judge(node->right, node->val, b);
// }

// bool isValidBST(struct TreeNode* root){
//     if (root == NULL) return true;
//     return judge(root, LONG_MIN, LONG_MAX);
// }
#define MAX_N 10005
int *num, cnt;

void pre_order(struct TreeNode *node) {
    if (node == NULL) return ;
    pre_order(node->left);
    num[cnt++] = node->val;
    pre_order(node->right);
    return ;
}

bool isValidBST(struct TreeNode* root){
    num = (int *)malloc(sizeof(int) * MAX_N);
    cnt = 0;
    pre_order(root);
    for (int i = 1; i < cnt; i++) {
        if (num[i - 1] >= num[i]) return false;
    }
    return true;
}



/**
 * C++
 */
class Solution {
public:
    void inorder(TreeNode* root, vector<int> &vec) {
        if (root == nullptr) return ;
        inorder(root->left, vec);
        vec.push_back(root->val);
        inorder(root->right, vec);
        return ;
    }

    bool isValidBST(TreeNode* root) {
        vector<int> vec;
        inorder(root, vec);
        for (int i = 1; i < vec.size(); i++) {
            if (vec[i] <= vec[i - 1]) return false;
        }
        return true;
    }
};

File: LeetCode/98验证二叉搜索树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	var sta []*TreeNode
	cur := root
	var pre *TreeNode
	for cur != nil || len(sta) > 0 {
		if cur != nil {
			sta = append(sta, cur)
			cur = cur.Left
		} else {
			cur = sta[len(sta)-1]
			sta = sta[:len(sta)-1]
			if pre != nil && pre.Val >= cur.Val {
				return false
			}
			pre = cur
			cur = cur.Right
		}
	}
	return true
}